drivers/misc/coresight/nrf_etr.c

```diff
@@ -0,0 +1,238 @@
+		    (__attribute__((__section__(                                              \
+			    LINKER_DT_NODE_REGION_NAME(DT_PHANDLE(node, memory_regions)))))), \
+		    ())
+
+static const uint32_t wsize_mask = DT_REG_SIZE(ETR_BUFFER_NODE) / sizeof(int) - 1;
+static const uint32_t wsize_inc = DT_REG_SIZE(ETR_BUFFER_NODE) / sizeof(int) - 1;
+
+static volatile bool tbm_full;
+static volatile uint32_t base_wr_idx;
+static uint32_t etr_rd_idx;
+static bool volatile use_async_uart;
+
+static struct k_sem uart_sem;
+static const struct device *uart_dev = DEVICE_DT_GET(UART_NODE);
+static uint32_t frame_buf0[CORESIGHT_TRACE_FRAME_SIZE32] MEMORY_SECTION(UART_NODE);
+static uint32_t frame_buf1[CORESIGHT_TRACE_FRAME_SIZE32] MEMORY_SECTION(UART_NODE);
+static uint32_t *frame_buf = frame_buf0;
+
+K_KERNEL_STACK_DEFINE(etr_stack, CONFIG_NRF_ETR_STACK_SIZE);
+static struct k_thread etr_thread;
+
+/** @brief Get write index.
+ *
+ * It is a non-wrapping 32 bit write index. To get actual index in the ETR buffer
+ * result must be masked by ETR buffer size mask.
+ */
+static uint32_t get_wr_idx(void)
+{
+	uint32_t cnt = nrfx_tbm_count_get();
+
+	if (tbm_full && (cnt < wsize_mask)) {
+		/* TBM full event is generated when max value is reached and not when
+		 * overflow occurs. We cannot increment base_wr_idx just after the
+		 * event but only when counter actually wraps.
+		 */
+		base_wr_idx += wsize_inc;
+		tbm_full = false;
+	}
+
+	return cnt + base_wr_idx;
+}
+
+/** @brief Get amount of pending data in ETR buffer. */
+static uint32_t pending_data(void)
+{
+	return get_wr_idx() - etr_rd_idx;
+}
+
+/** @brief Get current read index.
+ *
+ * Read index is not exact index in the ETR buffer. It does not wrap (32 bit word).
+ * So ETR read index is derived by masking the value by the ETR buffer size mask.
+ */
+static void rd_idx_inc(void)
+{
+	etr_rd_idx += CORESIGHT_TRACE_FRAME_SIZE32;
+}
+
+/** @brief Attempt to process data pending in the ETR circular buffer.
+ *
+ * Data is processed in 16 bytes packages. Each package is a STPv2 frame which
+ * contain data generated by STM stimulus ports.
+ *
+ */
+static void process(void)
+{
+	static const uint32_t *const etr_buf = (uint32_t *)(DT_REG_ADDR(ETR_BUFFER_NODE));
+	static uint32_t sync_cnt;
+	int err;
+	uint32_t pending;
+
+	/* If function is called in panic mode then it may interrupt ongoing
+	 * processing. This must be carefully handled as function decodes data
+	 * that must be synchronized. Losing synchronization results in failure.
+	 *
+	 * Special measures are taken to ensure proper synchronization when
+	 * processing is preempted by panic.
+	 *
+	 */
+	while ((pending = pending_data()) >= MIN_DATA) {
+		/* Do not read the data that has already been read but not yet processed. */
+		if (sync_cnt || (CONFIG_NRF_ETR_SYNC_PERIOD == 0)) {
+			sync_cnt--;
+			sys_cache_data_invd_range((void *)&etr_buf[etr_rd_idx & wsize_mask],
```

Comment:
I would check that the buffer is cache line aligned before cache is invalidated. If the condition is not met this code would result in data corruption.

Reply:
Isn't the implementation for `sys_cache_data_invd_range` just the arm core_cm7.h `SCB_InvalidateDCache_by_Addr()` call? This looks like it would properly invalidate the range even if the buffer is not aligned. It'd be more efficient to have it 32B aligned

Reply:
added `BUILD_ASSERT`s to check if etr buffer is aligned with cache line size and if size is cache aligned.. Invalidating is done on 16 bytes which is not cache line (32) but buffer is only read by the core so it should not be an issue.

Reply:
> Isn't the implementation for `sys_cache_data_invd_range` just the arm core_cm7.h `SCB_InvalidateDCache_by_Addr()` call? This looks like it would properly invalidate the range even if the buffer is not aligned. It'd be more efficient to have it 32B aligned

I was wrong, it is using `cache_nrf.c` and there the `_cache_range()`  function but even that does seem to do proper range invalidation and makes sure it will invalidate all affected cache lines